lightning-c-bindings/demo.cpp

```diff
@@ -0,0 +1,480 @@
+	LDK::BlockNotifier blocks1 = BlockNotifier_new(ChainWatchInterfaceUtil_as_ChainWatchInterface(&chain1));
+
+	LDKManyChannelMonitor mon1 {
+		.this_arg = &chain1,
+		.add_monitor = add_channel_monitor,
+		.update_monitor = update_channel_monitor,
+		.get_and_clear_pending_htlcs_updated = monitors_pending_htlcs_updated,
+		.free = NULL,
+	};
+
+	LDK::KeysManager keys1 = KeysManager_new(&node_seed, network, 0, 0);
+	LDK::KeysInterface keys_source1 = KeysManager_as_KeysInterface(&keys1);
+	LDKSecretKey node_secret1 = keys_source1->get_node_secret(keys_source1->this_arg);
+
+	LDK::UserConfig config1 = UserConfig_default();
+
+	LDK::ChannelManager cm1 = ChannelManager_new(network, fee_est, mon1, broadcast, logger1, KeysManager_as_KeysInterface(&keys1), config1, 0);
+	BlockNotifier_register_listener(&blocks1, ChannelManager_as_ChainListener(&cm1));
+
+	LDK::CVec_ChannelDetailsZ channels = ChannelManager_list_channels(&cm1);
+	assert(channels->datalen == 0);
+
+	LDK::NetGraphMsgHandler net_graph1 = NetGraphMsgHandler_new(ChainWatchInterfaceUtil_as_ChainWatchInterface(&chain1), logger1);
+
+	LDK::MessageHandler msg_handler1 = MessageHandler_new(ChannelManager_as_ChannelMessageHandler(&cm1), NetGraphMsgHandler_as_RoutingMessageHandler(&net_graph1));
+
+	LDKThirtyTwoBytes random_bytes = keys_source1->get_channel_id(keys_source1->this_arg);
+	LDK::PeerManager net1 = PeerManager_new(msg_handler1, node_secret1, &random_bytes.data, logger1);
+
+	// Demo getting a channel key and check that its returning real pubkeys:
+	LDK::ChannelKeys chan_keys1 = keys_source1->get_channel_keys(keys_source1->this_arg, false, 42);
+	chan_keys1->set_pubkeys(&chan_keys1); // Make sure pubkeys is defined
+	LDKPublicKey payment_point = ChannelPublicKeys_get_payment_point(&chan_keys1->pubkeys);
+	assert(memcmp(&payment_point, &null_pk, sizeof(null_pk)));
+
+	// Instantiate classes for node 2:
+
+	LDKLogger logger2 {
+		.this_arg = (void*)2,
+		.log = print_log,
+		.free = NULL,
+	};
+
+	LDK::ChainWatchInterfaceUtil chain2 = ChainWatchInterfaceUtil_new(network);
+	LDK::BlockNotifier blocks2 = BlockNotifier_new(ChainWatchInterfaceUtil_as_ChainWatchInterface(&chain2));
+
+	LDKManyChannelMonitor mon2 {
+		.this_arg = &chain2,
+		.add_monitor = add_channel_monitor,
+		.update_monitor = update_channel_monitor,
+		.get_and_clear_pending_htlcs_updated = monitors_pending_htlcs_updated,
+		.free = NULL,
+	};
+
+	memset(&node_seed, 1, 32);
+	LDK::KeysManager keys2 = KeysManager_new(&node_seed, network, 0, 0);
+	LDK::KeysInterface keys_source2 = KeysManager_as_KeysInterface(&keys2);
+	LDKSecretKey node_secret2 = keys_source2->get_node_secret(keys_source2->this_arg);
+
+	LDK::ChannelHandshakeConfig handshake_config2 = ChannelHandshakeConfig_default();
+	ChannelHandshakeConfig_set_minimum_depth(&handshake_config2, 2);
+	LDK::UserConfig config2 = UserConfig_default();
+	UserConfig_set_own_channel_config(&config2, handshake_config2);
```

Comment:
Any reason why the C versions of these functions are called directly rather than going through the C++ objects? Likewise for other classes like `LDK::ChannelManager`, `LDK::PeerManager`, and `LDK::BlockNotifier`.

Reply:
Same issue as https://github.com/rust-bitcoin/rust-lightning/pull/618#discussion_r475247659 - we don't currently know how to print C/C++ code at all.

Reply:
Is there a way to make the operator () tell the compiler that its really std::move?

Reply:
> Same issue as [#618 (comment)](https://github.com/rust-bitcoin/rust-lightning/pull/618#discussion_r475247659) - we don't currently know how to print C/C++ code at all.

Ah, I was confused thinking that overloading the `->` operator gave us access to the underlying Rust type. I was thinking this because for traits the C++ wrappers has a C function pointer for each method, which can be accessed using `->`.

Is there any use overloading the `->` operator for non-trait classes?

> Is there a way to make the operator () tell the compiler that its really std::move?

Not sure I follow. Was this supposed to be on a different comment?

Reply:
> Is there any use overloading the -> operator for non-trait classes?

Hmm, likely not for regular opaque structs. For enums and the mapped generic types (Result, Tuple, etc) it's useful, though.

>  Is there a way to make the operator () tell the compiler that its really std::move?

What I meant is that we currently use the `operator()` to capture the "move" logic, but `operator()` doesn't return a `&&`, it returns the object. If there were a way to make `operator()` only take std::move(this), then we could theoretically tell the compiler what's going on and get the same benefit. Maybe we can require std::move(X)(), though that's pretty verbose.

Reply:
Doh! That *was* meant for another comment. I hate github. That was intended for https://github.com/rust-bitcoin/rust-lightning/pull/618#discussion_r475247659